core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/common/accessor/gizmo/GizmoMemberAccessorImplementor.java

```diff
@@ -78,35 +79,55 @@ public String getName() {
-
-        GizmoMemberInfo memberInfo = new GizmoMemberInfo(member, annotationClass);
-        // ************************************************************************
-        // MemberAccessor methods
-        // ************************************************************************
-        createConstructor(classCreator, memberInfo);
-        createGetDeclaringClass(classCreator, memberInfo);
-        createGetType(classCreator, memberInfo);
-        createGetGenericType(classCreator);
-        createGetName(classCreator, memberInfo);
-        createGetSpeedNote(classCreator);
-        createSupportSetter(classCreator, memberInfo);
-        createExecuteGetter(classCreator, memberInfo);
-        createExecuteSetter(classCreator, memberInfo);
-        createGetAnnotation(classCreator);
+    public static void defineAccessorFor(String className, ClassOutput classOutput, GizmoMemberInfo memberInfo) {
+        Class<? extends AbstractGizmoMemberAccessor> superClass = getCorrectSuperclass(memberInfo);
+        try (ClassCreator classCreator = ClassCreator.builder()
+                .className(className)
+                .superClass(superClass)
+                .classOutput(classOutput)
+                .setFinal(true)
+                .build()) {
+            classCreator.getFieldCreator("genericType", Type.class)
+                    .setModifiers(Modifier.FINAL);
+            classCreator.getFieldCreator("annotatedElement", AnnotatedElement.class)
+                    .setModifiers(Modifier.FINAL);
+
+            // ************************************************************************
+            // MemberAccessor methods
+            // ************************************************************************
+            createConstructor(classCreator, memberInfo);
+            createGetDeclaringClass(classCreator, memberInfo);
+            createGetType(classCreator, memberInfo);
+            createGetGenericType(classCreator);
+            createGetName(classCreator, memberInfo);
+            createExecuteGetter(classCreator, memberInfo);
+            if (superClass == AbstractReadWriteGizmoMemberAccessor.class) {
+                createExecuteSetter(classCreator, memberInfo);
+            }
+            createGetAnnotation(classCreator);
+        }
+    }
+
+    private static Class<? extends AbstractGizmoMemberAccessor> getCorrectSuperclass(GizmoMemberInfo memberInfo) {
+        AtomicBoolean supportsSetter = new AtomicBoolean();
+        memberInfo.getDescriptor().whenIsMethod(method -> {
+            supportsSetter.set(memberInfo.getDescriptor().getSetter().isPresent());
+        });
+        memberInfo.getDescriptor().whenIsField(field -> {
+            supportsSetter.set(true);
+        });
+        if (supportsSetter.get()) {
+            return AbstractReadWriteGizmoMemberAccessor.class;
+        } else {
+            return AbstractReadOnlyGizmoMemberAccessor.class;
+        }
     }
```

Comment:
Not sure if this is the correct detection of whether or not we need setters, @Christopher-Chianelli .

Reply:
That is correct detection